Writable Json DOM - JsonNumber implementation and tests

kasiabulat opened the pull request:
I am working on adding writable Json DOM under `System.Text.Json`. It will be modifiable, dictionary-backed API to complement the readonly `JsonDocument`, with type hierarchy consisting of the base type `JsonNode`. Currently, I have implemented `JsonNumber` which is supposed to represent Json numeric value. Other derived types which we are planning to add include: JsonString, JsonBoolean, JsonArray and JsonObject.

More scenarios are available in this class:
https://github.com/kasiabulat/corefx/blob/master/src/System.Text.Json/tests/WritableJsonApiTests.cs
(also in files: WritableJsonApiTests.TestAccessingNestedMembers.cs, WritableJsonApiTests.TestModifying.cs and WritableJsonApiTests.TestTransformations.cs)

cc: @joperezr @bartonjs @ericstj @ahsonkhan @terrajobst @JamesNK

dnfclas commented:
[![CLA assistant check](https://cla.dotnetfoundation.org/pull/badge/not_signed)](https://cla.dotnetfoundation.org/dotnet/corefx?pullRequest=39917) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla.dotnetfoundation.org/dotnet/corefx?pullRequest=39917) before we can accept your contribution.<br/><br/>:x: kasiabulat [sign now](https://cla.dotnetfoundation.org/dotnet/corefx?pullRequest=39917)<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla.dotnetfoundation.org/check/dotnet/corefx?pullRequest=39917) it.</sub>

ericstj commented:
Can you make sure to have an issue tracking the API review?

kasiabulat commented:
#39922

kasiabulat commented:
I created an issue: #39922 and added a PR in my fork with specification: https://github.com/kasiabulat/corefx/pull/4.

joperezr commented:
We will want that spec checked-in to corefx main fork as well, so do you mind sending another PR to merge that here as well plz?

ahsonkhan commented on src/System.Text.Json/src/System/Text/Json/Node/JsonNumber.cs:
```diff
@@ -0,0 +1,610 @@
+        public static implicit operator JsonNumber(decimal value) => new JsonNumber(value);
+
+        /// <summary>
+        ///   Compares <paramref name="obj"/> to the numeric value of this instance. 
```
This doesn't seem accurate since 1.1e1 and 11 are equal (numerically), but I believe the following would return false:
```C#
var number = new JsonNumber("1.1e1")
object anotherForm = (object)new JsonNumber("11");
return number.Equals(anotherForm); // false?
```

kasiabulat replied:
Right, I will change it.

ahsonkhan commented on src/System.Text.Json/src/System/Text/Json/Node/JsonNumber.cs:
```diff
@@ -0,0 +1,610 @@
+        ///   <see langword="true"/> if the numeric value of this instance matches <paramref name="other"/>,
+        ///   <see langword="false"/> otherwise.
+        /// </returns>
+        public bool Equals(JsonNumber other) => _value == other._value;
```
What if I pass in `null` to this method?

kasiabulat replied:
I will fix it.

ahsonkhan commented on src/System.Text.Json/src/System/Text/Json/Node/JsonNumber.cs:
```diff
@@ -0,0 +1,610 @@
+        ///   <see langword="true"/> if the numeric value of instances matches,
+        ///   <see langword="false"/> otherwise.
+        /// </returns>
+        public static bool operator ==(JsonNumber left, JsonNumber right) => left._value == right._value;
```
Same question here.

```C#
var number = new JsonNumber("1");
if (number == null)
{
  return;
}
```

bartonjs replied:
I think conditional dereferencing (`left?._value == right?._value`) has the correct truth table.

ahsonkhan commented on src/System.Text.Json/tests/System.Text.Json.Tests.csproj:
```diff
@@ -101,9 +101,8 @@
     <Compile Include="Utf8JsonReaderTests.ValueTextEquals.cs" />
     <Compile Include="Utf8JsonWriterTests.cs" />
-  </ItemGroup>
-  <ItemGroup>
```
I would keep these item groups separate (since the other one has a src file being linked in)

ahsonkhan commented on src/System.Text.Json/tests/System.Text.Json.Tests.csproj:
```diff
@@ -101,9 +101,8 @@
-  </ItemGroup>
-  <ItemGroup>
     <Compile Include="..\src\System\Text\Json\BitStack.cs" Link="BitStack.cs" />
+    <Compile Include="WritableJsonApiTests.JsonNumberUnitTests.cs" />
```
nit: Move to other item group and sort alphabetically.

ahsonkhan commented on src/System.Text.Json/tests/WritableJsonApiTests.JsonNumberUnitTests.cs:
```diff
@@ -0,0 +1,473 @@
+
+namespace System.Text.Json
+{
+    public static partial class WritableJsonApiTests
```
I would rename this to JsonNodeTests instead.
Same with the filename.
```suggestion
    public static partial class JsonNodeTests
```

In this case:
JsonNodeTests.JsonNumber.cs

(UnitTests suffix is redundant, imo)

ahsonkhan commented on src/System.Text.Json/tests/WritableJsonApiTests.JsonNumberUnitTests.cs:
```diff
@@ -0,0 +1,473 @@
+
+using Xunit;
+
+namespace System.Text.Json
```
```suggestion
namespace System.Text.Json.Tests
```

ahsonkhan commented on src/System.Text.Json/tests/WritableJsonApiTests.JsonNumberUnitTests.cs:
```diff
@@ -0,0 +1,473 @@
+        [InlineData(0b_0110_1010)]
+        [InlineData("1e400")]
+        [InlineData("1e+100000002")]
+        [InlineData("184467440737095516150.184467440737095516150")]
```
Can we also add a test case with a really large natural number (without decimal)?

ahsonkhan commented on src/System.Text.Json/tests/WritableJsonApiTests.JsonNumberUnitTests.cs:
```diff
@@ -0,0 +1,473 @@
+    public static partial class WritableJsonApiTests
+    {
+        [Theory]
+        [InlineData(-456)]
```
We should have all of these cases but within "" as well.

ahsonkhan commented on src/System.Text.Json/tests/WritableJsonApiTests.JsonNumberUnitTests.cs:
```diff
@@ -0,0 +1,473 @@
+        public static void TestDefaultCtor()
+        {
+            var jsonNumber = new JsonNumber();
+            Assert.Equal(0, jsonNumber.GetInt32());
```
Add Asserts for all the other `.GetX` APIs here.

ahsonkhan commented on src/System.Text.Json/tests/WritableJsonApiTests.JsonNumberUnitTests.cs:
```diff
@@ -0,0 +1,473 @@
+        [Theory]
+        [InlineData(0)]
+        [InlineData(17)]
+        [InlineData(255)]
```
nit: replace with byte.minvalue, byte.maxvalue.

ahsonkhan commented on src/System.Text.Json/tests/WritableJsonApiTests.JsonNumberUnitTests.cs:
```diff
@@ -0,0 +1,473 @@
+        public static void TestByte(byte value)
+        {
+            var jsonNumber = new JsonNumber();
+            jsonNumber.SetByte(value);
```
Also test with the custom ctor (not just SetByte).

There are 4 ways to create a JsonNumber:
1) Default ctor + Set
2) Numeric type ctor
3) Implicit cast
4) String ctor

We should exercise all 4 for each of the `Test{NumericType}` test here. We can then probably get rid of `TestImplicitOperators` which is only testing a particular value.

Same feedback elsewhere.

kasiabulat replied:
So probably I should remove `TestJsonNumberConstructor` as well?

ahsonkhan replied:
Up to you. I would be fine with either. As long as some of the interesting cases from that are covered elsewhere. I did like how you used object/dynamic for the various test cases there.

ahsonkhan commented on src/System.Text.Json/tests/WritableJsonApiTests.JsonNumberUnitTests.cs:
```diff
@@ -0,0 +1,473 @@
+        [InlineData(17)]
+        [InlineData(short.MinValue)]
+        [InlineData(short.MaxValue)]
+        public static void TestShort(short value)
```
nit: Add a test case that is unique within byte.MaxValue and short.MaxValue (i.e. within the short range but outside the smaller numeric type's range). Say, 4_000, -4_000.

Could be even explicit about it: `byte.MaxValue + 1`, `sbyte.MinValue -1`

ahsonkhan commented on src/System.Text.Json/tests/WritableJsonApiTests.JsonNumberUnitTests.cs:
```diff
@@ -0,0 +1,473 @@
+        [Theory]
+        [InlineData(0)]
+        [InlineData(-17)]
+        [InlineData(17)]
```
nit: Similarly here/elsewhere, add something above short.MaxValue (like 100_000 and -100_000)

OR:
`short.MaxValue + 1`, `short.MinValue - 1`

ahsonkhan commented on src/System.Text.Json/tests/WritableJsonApiTests.JsonNumberUnitTests.cs:
```diff
@@ -0,0 +1,473 @@
+        [InlineData(-17)]
+        [InlineData(sbyte.MinValue)]
+        [InlineData(sbyte.MaxValue)]
+        public static void TestSbyte(sbyte value)
```
nit: SByte
```suggestion
        public static void TestSByte(sbyte value)
```

ahsonkhan commented on src/System.Text.Json/tests/WritableJsonApiTests.JsonNumberUnitTests.cs:
```diff
@@ -0,0 +1,473 @@
+        }
+
+        [Fact]
+        public static void TestDecimal()
```
Make this a theory like all the others, with different test values (max/min/0/etc.).

kasiabulat replied:
I planned to, but `decimal.MaxValue` doesn't seem to be a constant expression. I get an error: "CS0182 An attribute argument must be a constant expression, typeof expression or array creation expression of an attribute parameter type".

ahsonkhan replied:
Good point, it's a static readonly: https://docs.microsoft.com/en-us/dotnet/api/system.decimal.maxvalue?view=netframework-4.8

You can still make it a theory with MemberData instead of InlineData and use an IEunmerable<object[]> to define the test cases. For example:
https://github.com/dotnet/corefx/blob/93d7aa1c1737b6da29d04b78557215e18eb786d4/src/System.Text.Json/tests/Utf8JsonReaderTests.cs#L3201-L3203
https://github.com/dotnet/corefx/blob/93d7aa1c1737b6da29d04b78557215e18eb786d4/src/System.Text.Json/tests/Utf8JsonReaderTests.cs#L4448-L4459

As an aside, found some other interesting test cases:
https://github.com/dotnet/corefx/blob/93d7aa1c1737b6da29d04b78557215e18eb786d4/src/System.Text.Json/tests/Utf8JsonReaderTests.TryGet.cs#L689-L691

ahsonkhan commented on src/System.Text.Json/tests/WritableJsonApiTests.JsonNumberUnitTests.cs:
```diff
@@ -0,0 +1,473 @@
+        [InlineData("189")]
+        [InlineData("1e400")]
+        [InlineData("1e+100000002")]
+        [InlineData("184467440737095516150.184467440737095516150")]
```
Add test cases that may not roundtrip, like "1.1e1", "0.0", etc.

ahsonkhan commented on src/System.Text.Json/tests/WritableJsonApiTests.JsonNumberUnitTests.cs:
```diff
@@ -0,0 +1,473 @@
+        [InlineData("")]
+        [InlineData("3,14")]
+        [InlineData("this is not a number")]
+        [InlineData("NAN")]
```
Some more invalid cases: Hex values, comma separators, etc.

ahsonkhan commented on src/System.Text.Json/tests/WritableJsonApiTests.JsonNumberUnitTests.cs:
```diff
@@ -0,0 +1,473 @@
+        }
+
+        [Fact]
+        public static void TestIntFromString()
```
Why only int32? What about the others.

kasiabulat replied:
I will add others.

ahsonkhan commented on src/System.Text.Json/tests/WritableJsonApiTests.JsonNumberUnitTests.cs:
```diff
@@ -0,0 +1,473 @@
+            Assert.Equal(5, jsonNumber.GetInt32());
+
+            jsonNumber.SetDouble(3.14);
+            Assert.Equal(3.14, jsonNumber.GetDouble());
```
nit: Add Assert.Throws for jsonNumber.GetInt32() after the change of value.

kasiabulat replied:
I think it is already tested by `TestRationalGetMismatches`.

ahsonkhan commented on src/System.Text.Json/tests/WritableJsonApiTests.JsonNumberUnitTests.cs:
```diff
@@ -0,0 +1,473 @@
+            Assert.Equal((uint)127, jsonNumber.GetUInt32());
+
+            jsonNumber.SetFormattedValue("1e400");
+            Assert.Equal("1e400", jsonNumber.ToString());
```
What would `jsonNumber.GetDouble()` return in this case?

ahsonkhan commented on src/System.Text.Json/tests/WritableJsonApiTests.JsonNumberUnitTests.cs:
```diff
@@ -0,0 +1,473 @@
+        [InlineData(3.14)]
+        [InlineData(-15.5)]
+        [InlineData(double.MinValue)]
+        [InlineData(double.MaxValue)]
```
How about adding double.Infinity, -infinity, etc.? (same with float).

ahsonkhan commented on src/System.Text.Json/tests/WritableJsonApiTests.JsonNumberUnitTests.cs:
```diff
@@ -0,0 +1,473 @@
+        }
+
+        [Fact]
+        public static void TestUpcasts()
```
Will up-casting any JsonNumber always succeed? What about double <-> decimal. What's the up-cast relationship there? Or ulong <-> double?

ahsonkhan commented on src/System.Text.Json/tests/WritableJsonApiTests.JsonNumberUnitTests.cs:
```diff
@@ -0,0 +1,473 @@
+        }
+
+        [Fact]
+        public static void TestImplicitOperators()
```
Add tests for the equality (`.Equals(...)`, `==`, `!=`, and going through the `IEquatable<JsonNumber>` interface), including comparison with `null`.

Also for `GetHashCode()`.

As an aside Can you please measure the code coverage for the new code here and share that?
https://github.com/dotnet/corefx/blob/4b6562e0da17ba1d92d72bc2406af9eea3e9337a/Documentation/building/code-coverage.md
Running the following in the `corefx\src\System.Text.Json\tests\` directory should generate the report.
```text
dotnet msbuild /t:BuildAndTest /p:Coverage=true
```

kasiabulat commented on src/System.Text.Json/src/System/Text/Json/Node/JsonNode.cs:
```diff
@@ -0,0 +1,17 @@
+    /// </summary>
+    public abstract partial class JsonNode
+    {
+        private protected JsonNode() { }
```
Comment addressing it: https://github.com/kasiabulat/corefx/pull/2/files#r305451743

joperezr replied:
The idea of making it private protected was that only we (System.Text.Json) would be able to create sub-types of this.

bartonjs replied:
`internal` works, too; but we need a non-public ctor to stop the type from being extended outside the assembly (which won't make sense to JsonElement, at minimum)

ahsonkhan replied:
That's what I was thinking. Just make it internal.

kasiabulat commented on src/System.Text.Json/src/System/Text/Json/Node/JsonNumber.cs:
```diff
@@ -0,0 +1,610 @@
+        public JsonNumber(string value) => SetFormattedValue(value);
+
+        /// <summary>
+        ///   Initializes a new instance of the <see cref="JsonNumber"/> class from a Byte value.
```
Shouldn't I use programming language-neutral text?

bartonjs replied:
The `<see cref="byte"/>` here gets turned into `<see cref="T:System.Byte"/>` on the docs site, whch then will render as a link to the type with the text `Byte`.  It's indirect, but works out in the end :smile:

ahsonkhan replied:
As an example, see how it is getting rendered here (which is basically what you were initially looking for anyway):
https://docs.microsoft.com/en-us/dotnet/api/system.text.json.jsonelement.getbyte?view=netcore-3.0#System_Text_Json_JsonElement_GetByte

kasiabulat commented on src/System.Text.Json/src/System/Text/Json/Node/JsonNumber.cs:
```diff
@@ -0,0 +1,610 @@
+        ///   Converts the numeric value of this instance to its equivalent string representation.
+        /// </summary>
+        /// <returns>The string representation of the value of this instance.</returns>
+        public override string ToString() => _value;
```
I will add a remark, currently it is stored exactly as it was passed, so this will round-trip.

kasiabulat commented on src/System.Text.Json/src/System/Text/Json/Node/JsonNumber.cs:
```diff
@@ -0,0 +1,610 @@
+        {
+            if (value == null)
+                throw new ArgumentNullException(nameof(value));
+            if (value.Length == 0)
```
Ok, I'll change it, thanks!

kasiabulat commented on src/System.Text.Json/src/System/Text/Json/Node/JsonNumber.cs:
```diff
@@ -0,0 +1,610 @@
+            if (value.Length == 0)
+                throw new ArgumentException("Expected number, but instead got empty string.", nameof(value));
+            
+            JsonWriterHelper.ValidateNumber(Encoding.UTF8.GetBytes(value).AsSpan());
```
I created an issue: https://github.com/dotnet/corefx/issues/39957

joperezr commented on src/System.Text.Json/src/System.Text.Json.csproj:
```diff
@@ -22,7 +22,9 @@
     <Compile Include="System\Text\Json\JsonException.cs" />
     <Compile Include="System\Text\Json\JsonHelpers.cs" />
     <Compile Include="System\Text\Json\JsonHelpers.Date.cs" />
+    <Compile Include="System\Text\Json\Node\JsonNode.cs" />
```
I just noticed you added these files under Node subfolder. Traditionally we have tried to have the folder structure mimic the namespace structure as well as close as possible, but I do noticed that this project in particular isn't really following the convention. @ahsonkhan I guess it's really up to you if you want to remove the Node subfolder or not. If we do decide to keep the folder, I'm not sure if Node is the right name for it.

ahsonkhan replied:
This has come up before and we mainly deferred it to avoid churn/refactoring as we focused on stabilizing for 3.0 (and to avoid unnecessary merge conflicts/loss of history): https://github.com/dotnet/corefx/issues/38408#issuecomment-500544013

We can certainly clean up the file structure to be consistent after 3.0 ships.

Based on what we discussed offline, we should also consider moving the new JsonNode related files being added in this PR to its own `ItemGroup` (to keep master -> release/3.0 ports and cherry-picks of other, unrelated changes easier and without unnecessary merge conflict resolution).

joperezr commented on src/System.Text.Json/src/System/Text/Json/Node/JsonNumber.cs:
```diff
@@ -0,0 +1,610 @@
+    /// <summary>
+    ///   Represents JSON numeric value.
+    /// </summary>
+    public partial class JsonNumber : JsonNode, IEquatable<JsonNumber>
```
Same comment with partial that @ahsonkhan made on JsonNode

joperezr commented on src/System.Text.Json/src/System/Text/Json/Node/JsonNumber.cs:
```diff
@@ -0,0 +1,610 @@
+        /// </exception>
+        public void SetFormattedValue(string value)
+        {
+            if (value == null)
```
@ahsonkhan question: is System.Text.Json going to be annotated with nullability?

stephentoub replied:
> is System.Text.Json going to be annotated with nullability?

Eventually, yes, along with the rest of corefx.  For 3.0 we've only annotated a subset of the reference assemblies and corresponding implementations, namely those that have implementations in corelib.

joperezr commented on src/System.Text.Json/src/System/Text/Json/Node/JsonNumber.cs:
```diff
@@ -0,0 +1,610 @@
+        ///   <see langword="true"/> if the numeric value of this instance matches <paramref name="obj"/>,
+        ///   <see langword="false"/> otherwise.
+        /// </returns>
+        public override bool Equals(object obj) => obj is JsonNumber number && _value == number._value;
```
Is this right? I guess that this would mean that if I have a JsonNumber(0) and a JsonNumber("0.0") this would return false? same for JsonNumber(0.0) and JsonNumber("0.00").

ahsonkhan replied:
I brought up a similar concern since the API comment makes it seem like the values are mathematically equivalent: https://github.com/dotnet/corefx/pull/39917#discussion_r309510072

Given our underlying representation is string in the current design, numeric values that are equivalent mathematically will not be equal here.

Looks like Newtonsoft.Json's `JValue` does quite a sophisticated equality under the covers:
https://github.com/JamesNK/Newtonsoft.Json/blob/7b8c3b0ed0380cf76d66894e81bf4d4d5b0bd796/Src/Newtonsoft.Json/Linq/JValue.cs#L834
https://github.com/JamesNK/Newtonsoft.Json/blob/7b8c3b0ed0380cf76d66894e81bf4d4d5b0bd796/Src/Newtonsoft.Json/Linq/JValue.cs#L246

But it too treats these as not equal (since they mismatch on the type).

However, it does treat 1.1e1 and 1.10e1 as equal:
```C#
JObject o = JObject.Parse(@"{'v1': 0, 'v2': 0.0, 'v3': '0.0'}");
JToken tok1 = o["v1"];
JToken tok2 = o["v2"];
JToken tok3 = o["v3"];
Assert.Equal(tok1, tok2);
Assert.Equal(tok2, tok3);

JValue v1 = (JValue)o["v1"];
JValue v2 = (JValue)o["v2"];
JValue v3 = (JValue)o["v3"];
Assert.False(v1.Equals(v2)); // v1 = Type Integer
Assert.False(v2.Equals(v3)); // v2 = Type Float
Assert.False(v3.Equals(v1)); // v3  = Type String

o = JObject.Parse(@"{'v1': 11, 'v2': 1.1e1, 'v3': 1.10e1}");
v1 = (JValue)o["v1"];
v2 = (JValue)o["v2"];
v3 = (JValue)o["v3"];
Assert.False(v1.Equals(v2)); // v1 = Type Integer
Assert.True(v2.Equals(v3)); // v2 == v3 = Type Float
```

bartonjs commented on src/System.Text.Json/src/System/Text/Json/Node/JsonNumber.cs:
```diff
@@ -0,0 +1,610 @@
+        /// <exception cref="ArgumentException">
+        ///   Provided value is not in a legal JSON number format.
+        /// </exception>
+        public JsonNumber(string value) => SetFormattedValue(value);
```
Really the only thing missing from "legal JSON value" is scientific; which we could/should make work.

bartonjs commented on src/System.Text.Json/src/System/Text/Json/Node/JsonNumber.cs:
```diff
@@ -0,0 +1,614 @@
+        /// <exception cref="FormatException">
+        ///   <see cref="JsonNumber"/> represents a number in a format not compliant with <see cref="float"/>.
+        /// </exception>
+        public float GetSingle() => float.Parse(_value);
```
For scientific mode values this could be

```C#
public float GetSingle()
{
    if (float.TryParse(_value, out float value))
    {
        return value;
    }

    return float.Parse(_value, NumberStyles.AllowExponent);
}
```

(and similarly for double)

ahsonkhan replied:
The default float parser already supports `NumberStyles.AllowExponent` (and a lot more). In fact, specifying `NumberStyles.AllowExponent` limites what float.Parse accepts (it would fail to parse decimals for instance).

https://github.com/dotnet/corefx/blob/cc5653fcceefce5eac38dc7847b12b371027f78f/src/Common/src/CoreLib/System/Globalization/NumberStyles.cs#L61-L62
https://github.com/dotnet/corefx/blob/cc5653fcceefce5eac38dc7847b12b371027f78f/src/Common/src/CoreLib/System/Single.cs#L298-L302

Same with double.

https://github.com/dotnet/corefx/blob/cc5653fcceefce5eac38dc7847b12b371027f78f/src/Common/src/CoreLib/System/Double.cs#L299-L303

bartonjs commented on src/System.Text.Json/src/System/Text/Json/Node/JsonNumber.cs:
```diff
@@ -0,0 +1,610 @@
+            if (value == null)
+                throw new ArgumentNullException(nameof(value));
+            if (value.Length == 0)
+                throw new ArgumentException("Expected number, but instead got empty string.", nameof(value));
```
In .NET Framework localization was a concern.  So far, as far as I know, we're not localizing exception strings on .NET Core, but being in the Strings.resx file does mean that a) there aren't two different copies that get out of sync if the message gets tweaked and b) we could localize if desired.

ahsonkhan replied:
Let's fix this to use Strings.resx like elsewhere in this project.

bartonjs commented on src/System.Text.Json/tests/JsonNodeTests.JsonNumber.cs:
```diff
@@ -0,0 +1,632 @@
+
+namespace System.Text.Json.Tests
+{
+    public static partial class JsonNodeTests
```
I think I'd make this just JsonNumberTests, not partial on JsonNodeTests.  JsonNumber is testable in its own right, and you might want to just run all the JsonNumber tests (e..g `dotnet msbuild /t:BuildAndTest /p:XunitOptions="-class System.Text.Json.Tests.JsonNumberTests"`)

bartonjs commented on src/System.Text.Json/tests/JsonNodeTests.JsonNumber.cs:
```diff
@@ -0,0 +1,632 @@
+            Assert.Equal(value, jsonNumber.GetByte());
+            Assert.True(jsonNumber.TryGetByte(out result));
+            Assert.Equal(value, result);
+        }
```
What state are all of the other Get/TryGet methods supposed to be in? (If the "cross-type" tests are later, that's fine).

bartonjs commented on src/System.Text.Json/tests/JsonNodeTests.JsonNumber.cs:
```diff
@@ -0,0 +1,632 @@
+            jsonNumber.SetInt16(value);
+            Assert.Equal(value, jsonNumber.GetInt16());
+            Assert.True(jsonNumber.TryGetInt16(out short result));
+            Assert.Equal(value, result);
```
All of these tests are the same thing repeated 4 times, repeated N times.  Consider factoring.

```C#
private delegate bool TryGetValue<T>(out T result);

private static void SimpleNumberTests<T>(
        T value,
        Func<JsonNumber, T> ctor,
        Action<JsonNumber, T> setter,
        Func<T, JsonNumber> getter, TryGetValue<T> tryGetter)
{
    JsonNumber number = new JsonNumber();
    setter(number, value);
    AssertValue(value, number, getter, tryGetter);

    number = ctor(value);
    AssertValue(value, number, getter, tryGetter);

    number = new JsonNumber(value.ToString());
    AssertValue(value, number, getter, tryGetter);
}

private static void AssertValue<T>(
        T value,
        JsonNumber number,
        Func<T, JsonNumber> getter,
        TryGetValue<T> tryGetter)
{
    Assert.Equal(value, getter(number));
    Assert.True(tryGetter(number, out T result));
    Assert.Equal(value, result);
}

[Theory]
[InlineData(0)]
[InlineData(-17)]
[InlineData(17)]
[InlineData(short.MinValue)]
[InlineData(short.MaxValue)]
public static void TestShort(short value)
{
    SimpleNumberTests(
        value,
        v => new JsonNumber(v),
        (number, v) => number.SetInt16(v),
        number => number.GetInt16(),
        (number, out v) =>number.TryGetValue(out v));
}
```

or something like that.

It'd lose the implicit cast, but you could build those to call the same ultimate helper.

bartonjs replied:
(Actually, you should be able to do the implicit cast as another Func<JsonNumber, T>, something like `v => v`, which isn't quite the identity lambda it looks :smile:)

kasiabulat replied:
Thanks, that's exactly what I was missing!

bartonjs commented on src/System.Text.Json/tests/JsonNodeTests.JsonNumber.cs:
```diff
@@ -0,0 +1,632 @@
+
+        [Theory]
+        [InlineData(0)]
+        public static void TestDecimal(decimal value)
```
The theory on this one would probably need to take a `(string valueString)`, then immediately do `decimal value = decimal.Parse(valueString)`.   Alternatively use MemberData and make an array of literals using the `m` suffix.  e.g. (`3.141592658m`)

JamesNK commented:
What about supporting converting to and from BigInteger?

bartonjs commented:
> What about supporting converting to and from BigInteger?

There was pushback from that during the JsonReader API review, so I suggested it be left out for now.  `BigInteger.Parse(number.ToString())` and `new JsonNumber(bigInteger.ToString())` are both viable workarounds in the meantime.  (Part of why I got the String support in here :smile:)

JamesNK commented:
Json.NET doesn't actually expose BigInteger anywhere in its public API. JValue has a ctor that takes object so this is fine:

```cs
var i = new JValue(new BigInteger(100000000000000000000000000000));
```

What about this scenario:

```cs
var o = new JsonObject
{
    ["BigInt"] = new BigInteger(10000000000000000000000000000000000)
}
```

My point of view: Not supporting BigInteger on a low-level reader/writer API is fine, but these JSON DOM objects are going to be used heavily in peoples app code. Don't make them jump through hoops and have to search StackOverflow for work-arounds when they want to use BigInteger.

JamesNK commented:
As mentioned in a comment on the spec PR: Have you considered internal subtypes from JsonNumber? That will avoid the overhead of allocating a string for each value.

https://github.com/dotnet/corefx/pull/39954#discussion_r309889640

ahsonkhan commented on src/System.Text.Json/src/System/Text/Json/Node/JsonNumber.cs:
```diff
@@ -0,0 +1,640 @@
+        /// </exception>
+        /// <remarks> 
+        ///   On .NET Core this method does not return <see langword="false"/> for values larger than 
+        ///   <see cref="float.MaxValue"/> (or smaller than <see cref="float.MinValue"/>), 
```
double.MaxValue, not float. Same for the Try method
```suggestion
        ///   <see cref="double.MaxValue"/> (or smaller than <see cref="double.MinValue"/>), 
```

ahsonkhan commented on src/System.Text.Json/src/System/Text/Json/Node/JsonNumber.cs:
```diff
@@ -0,0 +1,640 @@
+        /// </summary>
+        /// <param name="obj">The object to compare against.</param>
+        /// <returns>
+        ///   <see langword="true"/> if the value of this instance matches exactly <paramref name="obj"/> (is equal and has the same format),
```
nit:
```suggestion
        ///   <see langword="true"/> if the value of this instance matches <paramref name="obj"/> exactly (is equal and has the same format),
```

ahsonkhan commented on src/System.Text.Json/src/System/Text/Json/Node/JsonNumber.cs:
```diff
@@ -0,0 +1,640 @@
+        /// </summary>
+        /// <param name="other">The JSON number to compare against.</param>
+        /// <returns>
+        ///   <see langword="true"/> if the value of this instance matches exactly <paramref name="other"/> (is equal and has the same format),
```
Same nit here:
nit:
```suggestion
        ///   <see langword="true"/> if the value of this instance matches <paramref name="other"/> exactly (is equal and has the same format),
```

ahsonkhan commented on src/System.Text.Json/tests/JsonNodeTests.JsonNumber.cs:
```diff
@@ -0,0 +1,534 @@
+
+namespace System.Text.Json.Tests
+{
+    public static class JsonNumberTests
```
nit: If we are not doing partial classes for JsonNode, rename the file to match the class name.

`JsonNumberTests.cs`

ahsonkhan commented on src/System.Text.Json/tests/JsonNodeTests.JsonNumber.cs:
```diff
@@ -0,0 +1,534 @@
+        [InlineData(0)]
+        [InlineData(-17)]
+        [InlineData(17)]
+        [InlineData(byte.MinValue-1)]
```
nit: spacing (same below in other tests)
```suggestion
        [InlineData(byte.MinValue - 1)]
```

ahsonkhan commented on src/System.Text.Json/tests/JsonNodeTests.JsonNumber.cs:
```diff
@@ -0,0 +1,534 @@
+            Assert.True(jsonNumber.Equals(new JsonNumber((ushort)123)));
+            Assert.True(jsonNumber.Equals(new JsonNumber("123")));
+            Assert.False(jsonNumber.Equals(new JsonNumber("123e1")));
+            Assert.False(jsonNumber.Equals(new JsonNumber(17)));
```
Add a test case for comparing `jsonNumber` to null and to an object.
```C#
            Assert.False(jsonNumber.Equals(null));

            object jsonNumberCopy = jsonNumber;
            object jsonNumberObject = new JsonNumber(123);
            Assert.True(jsonNumber.Equals(jsonNumberObject));
            Assert.True(jsonNumberCopy.Equals(jsonNumberObject));
            Assert.True(jsonNumberObject.Equals(jsonNumber));
```

ahsonkhan commented on src/System.Text.Json/tests/JsonNodeTests.JsonNumber.cs:
```diff
@@ -0,0 +1,534 @@
+            Assert.True(jsonNumber.Equals(new JsonNumber(123)));
+            Assert.True(jsonNumber.Equals(new JsonNumber((ushort)123)));
+            Assert.True(jsonNumber.Equals(new JsonNumber("123")));
+            Assert.False(jsonNumber.Equals(new JsonNumber("123e1")));
```
Add `123e0` as well (since 123e0 == 123 numerically)
```C#
            Assert.False(jsonNumber.Equals(new JsonNumber("123e0")));
```

ahsonkhan commented on src/System.Text.Json/tests/JsonNodeTests.JsonNumber.cs:
```diff
@@ -0,0 +1,534 @@
+        {
+            var jsonNumber = new JsonNumber(123);
+
+            Assert.True(jsonNumber.Equals(new JsonNumber(123)));
```
nit: Add commutative tests

So,
```C#
Assert.True(new JsonNumber(123).Equals(jsonNumber));
```

ahsonkhan commented on src/System.Text.Json/tests/JsonNodeTests.JsonNumber.cs:
```diff
@@ -0,0 +1,534 @@
+
+            IEquatable<JsonNumber> jsonNumberIEquatable = jsonNumber;
+            Assert.True(jsonNumberIEquatable.Equals(jsonNumber));
+            Assert.True(jsonNumber.Equals(jsonNumberIEquatable));
```
Let's add another Fact for `GetHashCode()` as well.

ahsonkhan commented on src/System.Text.Json/tests/JsonNodeTests.JsonNumber.cs:
```diff
@@ -0,0 +1,534 @@
+            Assert.Equal((ushort)0, jsonNumber.GetUInt16());
+            Assert.Equal((uint)0, jsonNumber.GetUInt32());
+            Assert.Equal((ulong)0, jsonNumber.GetUInt64());
+            Assert.Equal(0, jsonNumber.GetDecimal());
```
Also, `.GetHashCode()`, `.Equals(new JsonNumber())`, and `.Equals(new JsonNumber(5))`.

kasiabulat replied:
I'll add those to tests for Equals/GetHashCode.

ahsonkhan commented on src/System.Text.Json/tests/System.Text.Json.Tests.csproj:
```diff
@@ -101,6 +101,7 @@
     <Compile Include="Utf8JsonReaderTests.TryGet.Date.cs" />
     <Compile Include="Utf8JsonReaderTests.ValueTextEquals.cs" />
     <Compile Include="Utf8JsonWriterTests.cs" />
+    <Compile Include="JsonNodeTests.JsonNumber.cs" />
```
Based on previous discussion, let's move this to its own item group for now (both in source and test csproj). We can merge them back after 3.0.

kasiabulat commented:
I included all comments: added support for scientific notation and additional tests for rational types, equals and hash code. I fixed some errors from different architectures changing ',' to '.' where necessary and adding precision when comparing rationals.  I am still not sure if one test will pass -  getting previously set `double.MaxValue` seems to throw `System.OverflowException` on `Windows NETFX_x86_Release`. Do you know how can I fix it if it still fails?

Code coverage report for `JsonNumber` class:

```xml
<Summary 
  numSequencePoints="108"
  visitedSequencePoints="99"
  numBranchPoints="32"
  visitedBranchPoints="23"
  sequenceCoverage="91.66"
  branchCoverage="71.87"
  maxCyclomaticComplexity="4"
  minCyclomaticComplexity="1"
  visitedClasses="1"
  numClasses="1"
  visitedMethods="64"
  numMethods="65" />
```          

(result copied from `coverage.xml`, can I make msbuild to print for me coverage of selected class?)

kasiabulat commented:
I think `System.OverflowException` is not an issue anymore, but what would you suggest to use to solve comparing rationals problems?

ahsonkhan commented on src/System.Text.Json/src/System/Text/Json/Node/JsonNumber.cs:
```diff
@@ -0,0 +1,664 @@
+        /// <param name="value">The value to represent as a JSON number.</param>
+        public void SetSingle(float value)
+        {
+            if (value.Equals(float.PositiveInfinity) || value.Equals(float.NegativeInfinity))
```
Re-use existing validation logic for such checks.
https://github.com/dotnet/corefx/blob/cc5653fcceefce5eac38dc7847b12b371027f78f/src/System.Text.Json/src/System/Text/Json/Writer/JsonWriterHelper.cs#L55-L79

kasiabulat replied:
I wasn't aware of this, thanks!

ahsonkhan commented on src/System.Text.Json/src/System/Text/Json/Node/JsonNumber.cs:
```diff
@@ -0,0 +1,664 @@
+        {
+            if (value.Equals(float.PositiveInfinity) || value.Equals(float.NegativeInfinity))
+            {
+                throw new FormatException(SR.IllegalNumberFormat);
```
This should be ArgumentException.

Also, use existing exception message SpecialNumberValuesNotSupported.

ahsonkhan commented on src/System.Text.Json/src/Resources/Strings.resx:
```diff
@@ -423,4 +423,10 @@
+    <value>Expected number, but instead got empty string.</value>
+  </data>
+  <data name="IllegalNumberFormat" xml:space="preserve">
+    <value>Provided value does not represent legal JSON number.</value>
```
Use existing exception message `SpecialNumberValuesNotSupported`.

ahsonkhan commented on src/System.Text.Json/src/System/Text/Json/Node/JsonNumber.cs:
```diff
@@ -0,0 +1,664 @@
+        {
+            if(value.Equals(double.PositiveInfinity) || value.Equals(double.NegativeInfinity))
+            {
+                throw new FormatException(SR.IllegalNumberFormat);
```
Same here.

ahsonkhan commented on src/System.Text.Json/src/System/Text/Json/Node/JsonNumber.cs:
```diff
@@ -0,0 +1,664 @@
+        /// <summary>
+        ///   Changes the numeric value of this instance to represent a specified <see cref="double"/> value.
+        /// </summary>
+        /// <param name="value">The value to represent as a JSON number.</param>
```
Add exception xml tag in the comments to document what exceptions are thrown and when.

ahsonkhan commented on src/System.Text.Json/src/System/Text/Json/Node/JsonNumber.cs:
```diff
@@ -0,0 +1,664 @@
+        ///    Converts a <see cref="float"/> to a JSON number.
+        /// </summary>
+        /// <param name="value">The value to convert.</param>
+        public static implicit operator JsonNumber(float value) => new JsonNumber(value);
```
Hmm... this implicit cast can throw. Something worth thinking about since it goes against the FDG:
https://docs.microsoft.com/en-us/dotnet/standard/design-guidelines/operator-overloads#conversion-operators

```C#
JsonNumber number = float.PositiveInfinity; // throws
```

cc @bartonjs

ahsonkhan commented on src/System.Text.Json/tests/JsonNumberTests.cs:
```diff
@@ -0,0 +1,651 @@
+            {
+                double doubleValue => new JsonNumber(doubleValue.ToString(CultureInfo.InvariantCulture)),
+                float floatValue => new JsonNumber(floatValue.ToString(CultureInfo.InvariantCulture)),
+                decimal decimalValue => new JsonNumber(decimalValue.ToString(CultureInfo.InvariantCulture)),
```
A bunch of tests are failing because we don't use InvariantCulture when converting the numbers to string within the implementation (in both SetDouble and GetDouble, i.e. double -> string and double -> string), but are expecting the result to be culture invariant for some reason.

Why not leave it as default (using `CultureInfo.CurrentCulture`)?

```C#
double val = 3.14;
CultureInfo.CurrentCulture = new CultureInfo( "es-ES", false );
Console.WriteLine("CurrentCulture is now {0}.", CultureInfo.CurrentCulture.Name);
// when culture changed to spanish, period became comma in the output
Console.WriteLine(val.ToString()); // 3,14
Console.WriteLine(val.ToString(CultureInfo.InvariantCulture)); // 3.14

// 3.14 - since both sides use same culture (parse and ToString), this round trips fine
Console.WriteLine(double.Parse(val.ToString()));

// 314 - ToString was invariant, but parsing wasn't - so can't roundtrip (culture mismatch)
Console.WriteLine(double.Parse(val.ToString(CultureInfo.InvariantCulture))); 

// 3.14 - if both sides match, this round trips fine
Console.WriteLine(double.Parse(val.ToString(CultureInfo.InvariantCulture), CultureInfo.InvariantCulture));
```

bartonjs replied:
> Why not leave it as default (using CultureInfo.CurrentCulture)?

Because it then can't be sent to Utf8JsonWriter as a formatted number?

ahsonkhan replied:
My question was assuming the implementation will remain as is.

> Because it then can't be sent to Utf8JsonWriter as a formatted number?

In that case, we should change the underlying representation to be culture invariant (in the impl itself).

kasiabulat replied:
I changed the implementation to use Culture.Invariant.

kasiabulat commented on src/System.Text.Json/tests/JsonNumberTests.cs:
```diff
@@ -0,0 +1,651 @@
+        [InlineData(-17)]
+        [InlineData(3.14)]
+        [InlineData(-15.5)]
+#if BUILDING_INBOX_LIBRARY
```
When I run it locally for netfx ulong.Min works (because it's 0 and actually repeated test, so I will remove it at all), ulong.Max does not. Some of those throw `OverflowException` (double.X), some fail because of precision errors (`float.X`, `ulong.MaxValue`). But I am not sure if it make sense to include ifs in TestInitialization or write a new version of TestInitialization for just those few cases with AssertValue replaced wit AssertThrows. But case of double.MaxValue is actually tested in TestRationalGetMismatches, so I can add a platform-if there.

kasiabulat commented on src/System.Text.Json/tests/JsonNumberTests.cs:
```diff
@@ -0,0 +1,651 @@
+
+#if BUILDING_INBOX_LIBRARY
+        [Fact]
+        public static void TestScientificNotation()
```
Actually it works, sorry.

kasiabulat commented:
:tada: CI works! :tada:
There's only one thing left to consider - currently throwing implicit casts, but maybe for now I can just add it to open questions in specification. (?)

ericstj commented:
> but maybe for now I can just add it to open questions in specification. (?)

Or file an issue to track it, whichever you prefer.

stephentoub commented on src/System.Text.Json/src/System/Text/Json/Node/JsonNumber.cs:
```diff
@@ -0,0 +1,677 @@
+    /// </summary>
+    public class JsonNumber : JsonNode, IEquatable<JsonNumber>
+    {
+        private string _value;
```
This means that constructing a JsonNumber from a value requires converting it to a string?  That seems unnecessarily expensive.  Did we consider alternative representations?

kasiabulat replied:
We are thinking about changing it into one of the following: 
    -`Span<byte>` or array of `Utf8String`/`Char8` (once they come online in the future) / `byte`  
    - internal types that inherit from `JsonNumber` 
but we left the decision for API review.

ahsonkhan replied:
As @kasiabulat mentioned, we had discussed a few other alternatives and they are generally implementation detail specific. One of the design goals mentioned in the spec (https://github.com/dotnet/corefx/pull/39954) is

> Higher emphasis on usability over allocations/performance.

- Internal types that are specific to each numeric type in .NET with factories to create JsonNumber (@JamesNK suggested - https://github.com/dotnet/corefx/pull/39954#discussion_r309889640)
- Change to a internal struct field which has all the supported numeric types (something @terrajobst suggested that might be worth exploring).

Granted, I agree we should avoid any unnecessary overhead especially if there is no/little regression to usability. **We landed on keeping it as is for now, and revisit this once more of the feature is implemented, to unblock forward progress.**

@kasiabulat - can you please file an issue to explicitly track this? Or update the existing issue to talk about the underlying representation of a JsonNumber rather than just what `SetFormattedValue` does: https://github.com/dotnet/corefx/issues/39957

Another design point was about immutable vs mutable, but we landed on keeping it mutable.

stephentoub replied:
> we had discussed a few other alternatives

There are other approaches as well.  For example, rather than having a single string field, it could store:
```C#
private TypeCode _typeOf8bytes;
private ulong _8bytes;
private string _stringValue;
```
This would make the JsonNumber itself a little bigger, but it would allow any of the primitive types that are <= 8 bytes long (which is everything except for `decimal`) to be stored in the `_8bytes` field, with no additional allocation, formatting, or parsing required.  And as long as we expect the majority of `JsonNumber`s created to be created for `byte`s, `sbyte`s, `ushort`s, `short`s, `uint`s, `int`s, `ulong`s, `long`s, `float`s, or `double`s (rather than `decimal`s or `string`s), this would be seem to be a win.

> and they are generally implementation detail specific

Sure... but this PR is providing the implementation :)

> We landed on keeping it as is for now, and revisit this once more of the feature is implemented, to unblock forward progress.

Ok, as long as there's an issue tracking this and we'll revisit it.

ahsonkhan replied:
There was also some discussion on adding BigInteger support, but that would fall into the decimal/string category in the approach you suggested.

> primitive types that are <= 8 bytes long

Hmm, interesting. float/double can be represented as ulong without loss of precision?

kasiabulat replied:
(I edited the issue: https://github.com/dotnet/corefx/issues/39957)

stephentoub replied:
> float/double can be represented as ulong without loss of precision? 

Yes. Don't think of it as a long, think of it as an opaque 8 bytes that can be reinterpreted however you like, e.g. MemoryMarshal.Cast.

ahsonkhan replied:
> Don't think of it as a long, think of it as an opaque 8 bytes that can be reinterpreted however you like

Ah, makes sense (especially now that I look at the implementation of Single/Double more closely). Looks like each possible value in the set of floats/doubles can be uniquely represented by a unique 8-byte sequence.

https://docs.microsoft.com/en-us/dotnet/api/system.double?view=netframework-4.8#floating-point-representation-and-precision
The Double data type stores double-precision floating-point values in a **64-bit binary format**

ahsonkhan commented on src/System.Text.Json/tests/JsonNumberTests.cs:
```diff
@@ -0,0 +1,679 @@
+            // String constructor:
+            number = value switch
+            {
+                // Adding CultureInfo.InvariantCulture to support tests on plantforms where `,` is a  default decimal separator instead of `.`
```
nit:
```suggestion
                // Adding CultureInfo.InvariantCulture to support tests on platforms where `,` is a  default decimal separator instead of `.`
```

ahsonkhan commented on src/System.Text.Json/tests/JsonNumberTests.cs:
```diff
@@ -0,0 +1,679 @@
+            Assert.False(jsonNumberNull == jsonNumber);
+
+            Assert.True(jsonNumber != jsonNumberNull);
+            Assert.True(jsonNumberNull != jsonNumber);
```
Nice!

ahsonkhan commented on src/System.Text.Json/src/System/Text/Json/Node/JsonNumber.cs:
```diff
@@ -0,0 +1,677 @@
+        ///  <see langword="true"/> if instance was converted successfully; 
+        ///  otherwise, <see langword="false"/>
+        /// </returns>
+        public bool TryGetDecimal(out decimal value) => decimal.TryParse(_value, NumberStyles.Float, CultureInfo.InvariantCulture, out value);
```
I think using float in this case is incorrect. I wouldn't use any NumberStyle, or use the default for decimal, which is NumberSytles.Number (which includes allowtrailingsign, and allowthousands, but doesn't allowexponents).

Same with GetDecimal.

https://github.com/dotnet/corefx/blob/3cdd1f49a13e61060e24c8221ec381ce4cc3fc94/src/Common/src/CoreLib/System/Decimal.cs#L464-L468
https://github.com/dotnet/corefx/blob/3cdd1f49a13e61060e24c8221ec381ce4cc3fc94/src/Common/src/CoreLib/System/Globalization/NumberStyles.cs#L58-L59

Maybe add a test too.
```C#
Console.WriteLine(decimal.Parse("1,000"));
Console.WriteLine(decimal.Parse("1,000", NumberStyles.Number));
Console.WriteLine(decimal.Parse("1,000", NumberStyles.Float)); // Throws System.FormatException: Input string was not in a correct format.
```

kasiabulat replied:
I am afraid with `TryParse` I have to specify `NumberStyle`, but I will change it into `NumberStyle.Number` for decimal and add tests.

ahsonkhan commented on src/System.Text.Json/src/System/Text/Json/Node/JsonNumber.cs:
```diff
@@ -0,0 +1,677 @@
+        ///   instead <see langword="true"/> is returned and <see cref="float.PositiveInfinity"/> (or 
+        ///   <see cref="float.NegativeInfinity"/>) is emitted. 
+        /// </remarks> 
+        public bool TryGetDouble(out double value) => double.TryParse(_value, NumberStyles.Float, CultureInfo.InvariantCulture, out value);
```
nit: Specifying NumberStyles is unnecessary (since the default Parse uses that NumberStyles.Float anyway). Consider using the other overload (same elsewhere):
```suggestion
        public bool TryGetDouble(out double value) => double.TryParse(_value, CultureInfo.InvariantCulture, out value);
```

kasiabulat replied:
I planned to, but `double.TryParse` seems to have only overloads with `IFormatProvider`, which takes `NumerStyles` as well: https://docs.microsoft.com/en-us/dotnet/api/system.double.tryparse?view=netcore-3.0 (`double.Parse` have it though).

ahsonkhan replied:
Ah... Good point. I was looking at the `Parse` APIs and assumed TryParse would have similar overloads. That is an odd inconsistency. It's fine to leave it as is then.

ahsonkhan replied:
@stephentoub, though we can't fix this down-level, do you think its worth adding an issue to make the API overloads consistent on Parse/TryParse (for all the numeric types - decimal, float, double, etc.)?

ahsonkhan replied:
@kasiabulat, when specifying the NumberStyle for the Try* APIs, make sure to match the ones used by default for Double/Float (include `NumberStyles.AllowThousands`).

https://github.com/dotnet/corefx/blob/bf8945497724dbe7c69ce8aaf4a24ac4fffc1a72/src/Common/src/CoreLib/System/Double.cs#L299-L303

stephentoub replied:
> do you think its worth adding an issue to make the API overloads consistent on Parse/TryParse (for all the numeric types - decimal, float, double, etc.)?

What is the inconsistency?

ahsonkhan replied:
The `Parse` overloads let you pass in the `NumberStyles`, and `IFormatProvider` independently. So, the caller can call the overload with fewer arguments if they want one of the arguments to remain default.
The `TryParse` overloads force you to pass in a `NumberStyles` if all you want to change is `IFormatProvider` (and vice versa).

Parse overloads:
https://github.com/dotnet/corefx/blob/d3911035f2ba3eb5c44310342cc1d654e42aa316/src/System.Runtime/ref/System.Runtime.cs#L2201-L2205
TryParse overloads:
https://github.com/dotnet/corefx/blob/d3911035f2ba3eb5c44310342cc1d654e42aa316/src/System.Runtime/ref/System.Runtime.cs#L2226-L2229

bartonjs replied:
Why would we allow thousands (e.g. commas)?  IIRC JSON doesn't allow commas, so we should never been in a state where we needed to parse them.  We should use the overloads of non-Try Parse that aren't flexible, instead of making the Try be as flexible in a way that shouldn't matter.

ahsonkhan replied:
> so we should never been in a state where we needed to parse them

That's a good point. Even in the ctor that accepts string, we detect/reject numbers that aren't JSON compliant.

> We should use the overloads of non-Try Parse that aren't flexible, instead of making the Try be as flexible in a way that shouldn't matter.

The question is what NumberStyle/parameter fits best to only allow parsing of input that is valid JSON. Do we want `JsonNumber.TryGetDecimal` to return true for "1e1", even if `decimal.TryParse("1e1", out _)` returns false? If so, leaving it as `NumberStyles.Float` makes sense.

kasiabulat replied:
Is the final conclusion to keep `NumberStyles.Float` everywhere (as it is now, for float, double and decimal, in Get and TryGet)?

ahsonkhan replied:
Yep. The only other difference is trailing sign, which isn't valid in JSON either, so we should leave it as is (i.e. use `NumberStyles.Float`) for all three.
```C#
decimal.Parse("1000-"); // valid here, but not valid JSON number
```

ahsonkhan commented on src/System.Text.Json/ref/System.Text.Json.cs:
```diff
@@ -156,6 +156,97 @@ public abstract partial class JsonNamingPolicy
+    {
+        internal JsonNode() { }
+    }
+    public partial class JsonNumber : System.Text.Json.JsonNode, System.IEquatable<System.Text.Json.JsonNumber>
```
Similar to the other JsonNode types, seal JsonNumber as well: https://github.com/dotnet/corefx/pull/40107#discussion_r311718941

ahsonkhan commented on src/System.Text.Json/src/Resources/Strings.resx:
```diff
@@ -423,4 +423,10 @@
+  <data name="EmptyStringToInitializeNumber" xml:space="preserve">
+    <value>Expected a number, but instead got empty string.</value>
+  </data>
+  <data name="IllegalNumberFormat" xml:space="preserve">
```
This can be removed now (no callers).

ahsonkhan commented on src/System.Text.Json/tests/JsonNumberTests.cs:
```diff
@@ -0,0 +1,679 @@
+#if BUILDING_INBOX_LIBRARY
+        [InlineData(float.MinValue)]
+        [InlineData(float.MaxValue)]
+        [InlineData(ulong.MinValue)]
```
Remove ulong.MinValue as its duplicate of 0?

ahsonkhan commented on src/System.Text.Json/tests/JsonNumberTests.cs:
```diff
@@ -0,0 +1,679 @@
+        [InlineData(float.MinValue)]
+        [InlineData(float.MaxValue)]
+        [InlineData(ulong.MinValue)]
+        [InlineData(ulong.MaxValue)]
```
I think this should work on netfx too:
```C#
// True
Console.WriteLine((float)ulong.MaxValue == float.Parse(ulong.MaxValue.ToString(CultureInfo.InvariantCulture)));
```

ahsonkhan commented:
The MacOS test failure is unrelated (in System.Net.Http.Functional.Tests): https://github.com/dotnet/corefx/issues/40115
